### PR TITLE
OIDC: Structured claims support in config.

### PR DESCRIPTION
**Pull request by ondrejvelisek**
Config structure of file perun-oidc-scopes.properties has changed. It uses JSON now.

Config file has to be replaced when change is deployed. Example file is n devel  perun-oidc-scopes.properties.new

### CODE REVIEW

**Comment by ondrejvelisek** [comment]
Change regex for attrbute to `urn:perun:.*`
Remove empty lines

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java [hunk 1]
```diff
@@ -70,44 +80,88 @@ public Object process(PerunSession sess, String method, Deserializer params) thr
...
+		}
+
+		return userinfo;
+
+	}
 
-					try {
-						userinfo.put(claimName, sess.getPerun().getAttributesManager().getAttribute(sess, user, attrName).getValue());
 
+	private JsonNode dfs(PerunSession sess, User user, JsonNode node) throws InternalErrorException, WrongAttributeAssignmentException, PrivilegeException, UserNotExistsException {
```
Comment (licehammer) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: Please, rename this method.

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java [hunk 2]
```diff
@@ -70,44 +80,88 @@ public Object process(PerunSession sess, String method, Deserializer params) thr
...
+			if (node.isTextual()) {
+				Matcher m = attrRegex.matcher(node.getTextValue());
+				if (m.matches()) {
+
+					try {
+						Object value = sess.getPerun().getAttributesManager().getAttribute(sess, user, node.getTextValue()).getValue();
+						if (value != null) {
+							return new TextNode(value.toString());
+						} else {
+							log.info("Attribute "+node.getTextValue()+" defined in config "+CONFIG_FILE+" is not present for user "+user);
```
Comment (licehammer) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: How can you be sure, that the problem is in CONFIG_FILE?

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java [hunk 3]
```diff
@@ -29,14 +35,18 @@
 
 	private final String USERINFO_METHOD = "userinfo";
 	private final String USERINFO_FOR_USER_METHOD = "getSpecificUserinfo";
-	private final String PROPERTIES_FILE = "perun-oidc-scopes.properties";
+	private final String CONFIG_FILE = "perun-oidc-scopes.properties";
+
+	// if config value matches this regex it is consider as perun attribute and is replaced by real value.
+	private Pattern attrRegex = Pattern.compile("urn:perun:.*");
```
Comment (licehammer) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: "^urn:perun:.*" regex would be slightly better.

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java [hunk 4]
```diff
@@ -70,44 +80,88 @@ public Object process(PerunSession sess, String method, Deserializer params) thr
...
+		if (!node.isObject()) {
+
+			if (node.isTextual()) {
+				Matcher m = attrRegex.matcher(node.getTextValue());
+				if (m.matches()) {
+
+					try {
+						Object value = sess.getPerun().getAttributesManager().getAttribute(sess, user, node.getTextValue()).getValue();
+						if (value != null) {
+							return new TextNode(value.toString());
```
Comment (licehammer) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: Are you sure you want to get attribute value like this even if the value type is List or Map?

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java (hunk 2 above)
Comment (ondrejvelisek) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: I think text does not say problem is there. But attribute is configured there and is not set for the user.

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java (hunk 4 above)
Comment (ondrejvelisek) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: Well, you are right. I should convert it to JSON.

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java (hunk 3 above)
Comment (ondrejvelisek) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: OK

**Comment by ondrejvelisek** [comment]
I would say 'i' for Iterator is quite standard, isn't it?

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java (hunk 2 above)
Comment (licehammer) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: I tried to point out that log msg claims it get something form CONFIG_FILE, but in reality data are obtained from argument of the method (therefore they don't have to originate in CONFIG_FILE).

**Comment by licehammer** [comment]
> I would say 'i' for Iterator is quite standard, isn't it?

Not in java :-)

To be honest, I don't mind if "cryptic name" is used in very small piece of code (e.g. one-liner). If you have it in more complex code or if you start using such names for more variable it's gets messy. 

For example if you see line like this:
```
String res = someMethod(i.next().getValue())
```
You have no idea what it is about unless you backtrack in the code to find out meaning of variables.

File: perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java (hunk 2 above)
Comment (ondrejvelisek) [inline perun-oidc/src/main/java/cz/metacentrum/perun/oidc/OIDC.java]: You are right.

**Comment by ondrejvelisek** [comment]
Ok. I think all of your change requests are in place now.

**Pull request merged**